Turn on tablets on keyspace by default when the feature is enabled

xemul opened the pull request:
To enable tablets replication one needs to turn on the (experimental) feature and specify the `initial_tablets: N` option when creating a keyspace. We want tablets to become default in the future and allow users to explicitly opt it out if they want to.

This PR solves this by changing the CREATE KEYSPACE syntax wrt tablets options. Now there's a new TABLETS options map and the usage is

* `CREATE KEYSPACE ...` will turn tablets on or off based on cluster feature being enabled/disabled
* `CREATE KEYSPACE ... WITH TABLETS = { 'enabled': false }` will turn tablets off regardless of what
* `CREATE KEYSPACE ... WITH TABLETS = { 'enabled': true }` will try to enable tablets with default configuration
* `CREATE KEYSPACE ... WITH TABLETS = { 'initial': <int> }` is now the replacement for `REPLICATION = { ... 'initial_tablets': <int> }` thing

fixes: #16319

avikivity commented on cql3/statements/create_keyspace_statement.cc:
```diff
@@ -92,11 +92,12 @@ void create_keyspace_statement::validate(query_processor& qp, const service::cli
 
     try {
-        m = service::prepare_new_keyspace_announcement(qp.db().real_database(), _attrs->as_ks_metadata(_name, tm), ts);
+        m = service::prepare_new_keyspace_announcement(qp.db().real_database(), _attrs->as_ks_metadata(_name, tm, feat), ts);
```
Theological modularity note: feature_service is anti-modular, since it has fan-in from many modules (anything that changes over time) and fan-out to many modules (anything that depends on a feature).

We should have broken in down at high level and just sent individual features to anyone that needs them. So main.cc sends the tablets feature to query_processor, not the whole thing.

This is just for consideration, I don't expect any action.

avikivity reviewed (approved):
Looks good.

Note: before we de-experimental-ize the tablets feature, we have to add a warning to CREATE KEYSPACE statements that turn on tablets to warn against missing features. But as long as it's still experimental, the user is protected (if disabled) or warned (if they enabled the feature).

avikivity commented:
Oh, please update docs/cql/ddl.rst.

tgrabiec commented on cql3/statements/ks_prop_defs.cc:
```diff
@@ -68,6 +68,39 @@ static std::map<sstring, sstring> prepare_options(
+        }
+
+        options.emplace(ks_prop_defs::INITIAL_TABLETS_KEY, fmt::to_string(initial_tablets));
+    }
```
It would be better to decide this on tablet allocation time, when table is created (in network_topology_strategy::allocate_tablets_for_new_table). The topology may have changed since the keyspace was created.

at keyspace creation time, tablet_aware_replication_strategy::process_tablet_options() should just set _uses_tablets = true when the feature is enabled.

I think we shouldn't set "initial_tablets" on behalf of the user.

xemul replied:
Not that I disagree, but the same applies to per-DC replication factors and they _are_ set here if not supplied by user. Why are tablets different in this sense?

tgrabiec replied:
With `initial_tablets` it's more harmful, since the user's intent is not to use a specific tablet count, while with RF he specified the intent to use a given `replication_factor` for each DC.

As Avi noted elsewhere, the tablet count should be actually specified per table. It's only per keyspace now because we need whole keyspace to be marked as tablet-based somehow. So want we really want to set is the "tablets: true" option, not "initial_tablets". 

The optimal value of initial tablets chosen by the system depends on current topology when tablets are allocated, and tablets are allocated not when the keyspace is created, but when a table is created. So we should choose it when creating a table.

xemul replied:
The above is in e.g. scylla-driver-3.26.0.zip/cassandra/metadata.py

```python
class ReplicationFactor(object):
    ...
    @staticmethod
    def create(rf):
        """
        Given the inputted replication factor string, parse and return the ReplicationFactor instance.
        """
        transient_replicas = None
        try:
            all_replicas = int(rf)
        except ValueError:
            try:
                rf = rf.split('/')
                all_replicas, transient_replicas = int(rf[0]), int(rf[1])
            except Exception:
                raise ValueError("Unable to determine replication factor from: {}".format(rf))

        return ReplicationFactor(all_replicas, transient_replicas)

```

nyh replied:
Yes, both "tablets" and the previous "initial-tablets" will be handled as if they are data center names if you stick them inside the topology strategy. Maybe it can be outside?
Alternatively, we can use "tablets: 0" and "tablets: 1". Ugly, but will work :-( Just not if you have a data center called "tablets"...

xemul replied:
Maybe it's better to have explicit dedicated replication strategy name for tablets?
And an optional quirk in prepare_options() that would overwrite NetworkTopologyReplicationStrategy with TabletsReplicationStrategy

xemul replied:
> Just not if you have a data center called "tablets"...

... or "initial_tablets" :)

tgrabiec replied:
> Maybe it's better to have explicit dedicated replication strategy name for tablets? And an optional quirk in prepare_options() that would overwrite NetworkTopologyReplicationStrategy with TabletsReplicationStrategy

That doesn't seem to save us work, since then you need to teach drivers about this new replication strategy. You can as well teach the drivers about tablet options.

Another reasons not to introduce separate RS for tablets is that tablets is orthogonal to RS spirit, you could make other strategies tablet-based too.

avikivity replied:
Alternatives:

 - a new WITH tablets = boolean_constant clause
 - use `$` - `tablets$enabled`, `tablets$initial_count` (the latter will go away, it belongs on the table level).

tgrabiec replied:
> Alternatives:
> 
> * a new WITH tablets = boolean_constant clause

That's cleaner, but it means we need to extend schema tables. This means we need to update tools which work with schema sstables. Did we phase-out java tools already?

> * use `$` - `tablets$enabled`, `tablets$initial_count` (the latter will go away, it belongs on the table level).

xemul replied:
Can it _not_ be the part of DDL? In that case we can put the bit into `system_schema.scylla_keyspaces`, this would solve the [boot-time schema merging issue](https://github.com/scylladb/scylladb/pull/16364#discussion_r1425388543)

tgrabiec commented on cql3/statements/ks_prop_defs.cc:
```diff
@@ -68,6 +68,39 @@ static std::map<sstring, sstring> prepare_options(
+            }
+
+            unsigned tablets_in_dc = (shards_in_dc + rf_in_dc - 1) / rf_in_dc;
+            initial_tablets = std::max(initial_tablets, tablets_in_dc);
```
initial_tablets must be a power of two.

xemul replied:
This is not checked currently. If user does `CREATE TABLE ... initial_tablets=13` it would accept that and apply "power of 2" later on. Why would automatic calculation be different?

tgrabiec replied:
Right, so it's fine.

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/16364)

**Changed Files:**
- [cql/ddl](http://previews.docs.scylladb.com/scylladb/16364/cql/ddl)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

avikivity commented on docs/cql/ddl.rst:
```diff
@@ -165,8 +165,16 @@ sub-option                             type  description
+``'initial_tablets'``                  int   The number of tablets each table starts with.
+                                             Applies only if `tablets=true`. If missing,
+                                             the number is selected automatically based on
+                                             datacenter refplication factors
```
I tried to document initial_tablets in https://github.com/scylladb/scylladb/pull/16193

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5162/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6552/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5162/)
#### Failed Tests (12/22265):
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [topology.test_topology_failure_recovery.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_topology_failure_recovery_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_topology_failure_recovery.debug.1)
- [topology.test_topology_failure_recovery.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_topology_failure_recovery_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_topology_failure_recovery.debug.2)
- [topology.test_topology_failure_recovery.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_topology_failure_recovery_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_topology_failure_recovery.debug.3)
- [topology.test_topology_failure_recovery.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5162/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_topology_failure_recovery_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_topology_failure_recovery.dev.1)
#### Build Details:
- Duration: 2 hr 25 min
- Builder: spider5.cloudius-systems.com

nyh commented:
This PR conflicts with https://github.com/scylladb/scylladb/pull/16353, which added optional (non-default) tablets supports for Alternator, and passed "initial_tablets" to the underlying keyspace creation function, which after this patch is no longer the right thing to do.

If you commit **this** PR first, I will need to redo my PR to fit the new mechanism.

xemul commented:
> This PR conflicts with #16353, which added optional (non-default) tablets supports for Alternator, and passed "initial_tablets" to the underlying keyspace creation function, which after this patch is no longer the right thing to do.

After this patch it's still possible to pass the `initial_tables` into underlying keyspace creation. The only difference is that this option becomes an _option_, i.e. -- when not specified, its value will be auto-estimated, but tables will be turned on anyway

xemul commented:
upd:
- drop doc part back in favor of #16193 
- do not introduce `tablets: true|false` option as it breaks external tools that treat the option as yet another dc replication factor
- estimate default initial tablets when a table is created

tgrabiec commented on locator/tablets.cc:
```diff
@@ -537,14 +537,18 @@ void tablet_aware_replication_strategy::validate_tablet_options(const gms::featu
-    auto i = opts.find(cql3::statements::ks_prop_defs::INITIAL_TABLETS_KEY);
-    if (i != opts.end()) {
-        _initial_tablets = parse_initial_tablets(i->second);
+    if (fs.tablets) {
```
I think we should record this decision in the schema somehow to be safe.

Otherwise, during upgrade, one node may still think the feature is disabled and anther node may already see that the feature is enabled, and they'll disagree about whether the keyspace is using tablets.

xemul replied:
> I think we should record this decision in the schema somehow to be safe.

Would record(s) in system.tablets work?

> Otherwise, during upgrade, one node may still think the feature is disabled and anther node may already see that the feature is enabled, and they'll disagree about whether the keyspace is using tablets.

If a node tries to join the cluster with tablets enabled it will need to enable tablets too and not turn it off since then, so presumably you're talking about some early start-up code? Where can this misunderstanding currently happen?

tgrabiec replied:
> > I think we should record this decision in the schema somehow to be safe.
> 
> Would record(s) in system.tablets work?

No, because partitions there are per-table not per keyspace.

> 
> > Otherwise, during upgrade, one node may still think the feature is disabled and anther node may already see that the feature is enabled, and they'll disagree about whether the keyspace is using tablets.
> 
> If a node tries to join the cluster with tablets enabled it will need to enable tablets too and not turn it off since then, so presumably you're talking about some early start-up code? Where can this misunderstanding currently happen?

I'm thinking about rolling upgrade which will enable tablets feature. There is a period where all nodes support the feature but not all of them enabled it.

xemul replied:
Actually current bootstrap code is already racy in this sense. Here's what I see from logs (some lines skipped):

```
16:24:05,391 storage_service - Performing gossip shadow round
16:24:05,391 gossip - Gossip shadow round started with nodes={127.83.115.13, 127.83.115.11, 127.83.115.7}
16:24:05,392 gossip - Gossip shadow round finished with nodes_talked={127.83.115.13, 127.83.115.7, 127.83.115.11}
16:24:05,392 storage_service - Checking bootstrapping/leaving/moving nodes: ok (check_for_endpoint_collision)
16:24:05,392 storage_service - Save advertised features list in the 'system.local' table
16:24:05,395 schema_tables - Schema version changed to 59adb24e-f3cd-3e02-97f0-5b395827453f
16:24:05,395 storage_service - Starting up server gossip
16:24:05,396 gossip - failure_detector_loop: Started main loop
16:24:05,396 raft_group0 - setup_group0: joining group 0...
16:24:05,396 raft_group0 - server c60209cb-28d4-4743-b59a-9e8602196367 found no local group 0. Discovering...
16:24:05,396 raft_group0 - server c60209cb-28d4-4743-b59a-9e8602196367 found group 0 with group id df10a1c0-99ba-11ee-9bf5-c955f5c66b4e, leader c80d3edd-fe43-41b6-ae33-8930afa41a4a
16:24:05,396 storage_service - raft topology: join: sending the join request to 127.83.115.11
16:24:05,415 storage_service - raft topology: join: request to join placed, waiting for the response from the topology coordinator
16:24:05,416 raft_group0 - Server c60209cb-28d4-4743-b59a-9e8602196367 is starting group 0 with id df10a1c0-99ba-11ee-9bf5-c955f5c66b4e
16:24:06,464 schema_tables - Creating keyspace system_auth
16:24:06,464 schema_tables - Creating keyspace system_distributed
16:24:06,464 schema_tables - Creating keyspace system_distributed_everywhere
16:24:06,464 schema_tables - Creating keyspace system_traces
16:24:06,464 schema_tables - Creating keyspace test
... <more "Creating keyspace ..." messages>
16:24:06,473 schema_tables - Tablet metadata changed
16:24:06,483 migration_manager - Gossiping my schema version e2c39868-99ba-11ee-47eb-96ff97452461
16:24:06,483 schema_tables - Schema version changed to e2c39868-99ba-11ee-47eb-96ff97452461
16:24:06,488 features - Feature TABLE_DIGEST_INSENSITIVE_TO_EXPIRY is enabled
...
16:24:06,535 features - Feature TABLETS is enabled
```

Note the sub-sequence:

```
16:24:06,464 schema_tables - Creating keyspace test
16:24:06,535 features - Feature TABLETS is enabled
```

So features are enabled in parallel (?) with merging schema and not seeing a feature when creating keyspace locally is becoming fatal.

tgrabiec replied:
It used to be the case that gossip enabled features, so it was in the background. Now it's enabled through raft, so it should be easy to serialize it with schema changes and make sure it's committed before bootstrap completes. \cc @piodul How hard is that?

xemul replied:
Features are enabled in several iterations

```
INFO  2023-12-13 17:53:21,445 [shard 0:strm] features - Enabling feature list [{}] from topology state load
INFO  2023-12-13 17:53:21,445 [shard 0:strm] features - Enabling feature list finished
INFO  2023-12-13 17:53:22,502 [shard 0:stmt] schema_tables - do_merge_schema -> start
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace system_auth
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace system_distributed
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace system_distributed_everywhere
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace system_traces
INFO  2023-12-13 17:53:22,510 [shard 0:stmt] schema_tables - Creating keyspace test
... <more "Creating ..." messages>
INFO  2023-12-13 17:53:22,520 [shard 0:stmt] schema_tables - Tablet metadata changed
INFO  2023-12-13 17:53:22,527 [shard 0:stmt] schema_tables - do_merge_schema <- done
INFO  2023-12-13 17:53:22,539 [shard 0:strm] features - Enabling feature list [{AGGREGATE_STORAGE_OPTIONS, ALTERNATOR_TTL, CDC, CDC_GENERATIONS_V2, COLLECTION_INDEXING, COMPUTED_COLUMNS, CORRECT_COUNTER_ORDER, CORRECT_IDX_TOKEN_IN_SECOND>
...
INFO  2023-12-13 17:53:22,591 [shard 0:strm] features - Feature TABLETS is enabled
...
INFO  2023-12-13 17:53:22,591 [shard 0:strm] features - Enabling feature list finished
```

the "Enabling feature list ... from topology state load" is the call to feature_service::enable() from storage_service::topology_state_load()

xemul replied:
Both feature-enabling happen from topology_sate_load() too and the "enabled_features" list is read from system.topology. 1st time it's empty, 2nd time it's full

xemul replied:
... and the update does _not_ happen via `topology_mutation_builder::add_enabled_features()`

xemul replied:
Features appear in system.topology while doing `transfer_snapshot()` that brings new schema to the node. So the steps are

* load_snapshot()
   - load_topology_state()
      - enable_feature( empty )
* transfer_snapshot()
   - merge_schema
      - create keyspaces
   - merge_topology_snapshot()
      - update system.topology.enabled_features
* load_snapshot() (another one?)
   - load_topology_state()
      - enable_feature( full )

piodul replied:
> So features are enabled in parallel (?) with merging schema and not seeing a feature when creating keyspace locally is becoming fatal.

@xemul In the logs you attached, it looks like a group0 snapshot transfer can be seen. In the snapshot transfer code we merge the schema first and only then call topology_state_load (where we mark features as enabled in feature_service).

\cc @piodul How hard is that?

@tgrabiec Both enabling a feature and modifying schema happens via a group0 command. If they are applied one after another then you are safe, but you may get a snapshot transfer. Perhaps we could change it to apply schema after features are marked as enabled, but @gleb-cloudius will most likely not like introducing such dependencies (my first design of cluster feature on raft did something like that and was rejected).

Besides, I see some other reasons not to do it like that:

- It's not clear to me whether there will be an issue in recovery. In recovery, we don't use raft to manage schema or cluster features.
- Relying on the property that "feature is enabled on this node => other node will see it as enabled at the right time" requires some non-local reasoning and knowing that the code is always called from group0 command application / snapshot transfer. As you mentioned, in general feature being enabled locally only implies that other nodes _support it_, but implies nothing about being seen as enabled.

xemul replied:
> @xemul In the logs you attached, it looks like a group0 snapshot transfer can be seen. In the snapshot transfer code we merge the schema first and only then call topology_state_load (where we mark features as enabled in feature_service).

Yes, I already [discovered](https://github.com/scylladb/scylladb/pull/16364#discussion_r1425664697) that

tgrabiec replied:
> * load_snapshot()
>   
>   * load_topology_state()
>     
>     * enable_feature( empty )
> * transfer_snapshot()
>   
>   * merge_schema
>     
>     * create keyspaces
>   * merge_topology_snapshot()
>     
>     * update system.topology.enabled_features
> * load_snapshot() (another one?)
>   
>   * load_topology_state()
>     
>     * enable_feature( full )

It seems wrong that features are enabled post snapshot transfer (esp. schema), since snapshot content may depend on those features.

gleb-cloudius replied:
> It seems wrong that features are enabled post snapshot transfer (esp. schema), since snapshot content may depend on those features.

The node suppose to be able to handle all features is supports even if they are not enabled yet. It only starts producing output that depends on the feature when it becomes enabled. This is the assumption with features over gossiper as well. Since the snapshot contains the feature as enabled the nodes that receives it has to support it already, so everything is fine.

xemul replied:
> Since the snapshot contains the feature as enabled the nodes that receives it has to support it already, so everything is fine.

Node does _support_ it already, the problem is that code than wants to see the tablets feature as _enabled_ is run first, then enabled features from the snapshot are enabled in the feature service

gleb-cloudius replied:
> > Since the snapshot contains the feature as enabled the nodes that receives it has to support it already, so everything is fine.
> 
> Node does _support_ it already, the problem is that code than wants to see the tablets feature as _enabled_ is run first, then enabled features from the snapshot are enabled in the feature service

Why does the code cares if the feature is enabled in _this_ node? If it gets something tablet related from the other node it should be able to process it since this can only happen if the node supports the feature. It should not do `if (tablets not enabled yet) do something bad`. This is the same with gossiper IIRC. There is no guaranty all nodes see feature entanglements at the same time, what is guarantied is that if one node starts using the feature all nodes support it.

xemul replied:
> Why does the code cares if the feature is enabled in this node?

Because when we check features we do it like `if ((bool)feature_service.local().feature_name) { /* do something when feature is enabled */ } In particular, when a new node joins the cluster it merges schema, when schema is merged keyspaces are created, when a keyspace is created it creates replication strategy and if r.s. is the network replication strategy, its constructor checks for feature_service.tablets to be true/false (i.e. -- the feature is enabled/disabled) and then sets "per-table" bit accordingly. Not seeing the feature enabled early enough, results in vnode-based network replication strategy which is wrong, since all other nodes already see this keyspace as per-table r.s.

gleb-cloudius replied:
>  its constructor checks for feature_service.tablets to be true/false (i.e. -- the feature is enabled/disabled) and then sets "per-table" bit accordingly.

I looks like the "per-table" bit  should be part of the schema, not decided locally based on incomplete info.

xemul replied:
> I looks like the "per-table" bit should be part of the schema, not decided locally based on incomplete info.

OK. Any place you have in mind? Part of `with replication = { ... }` set doesn't work, all field in it [are treated as replication factor integers](https://github.com/scylladb/scylladb/pull/16364#discussion_r1422569803)

gleb-cloudius replied:
> > I looks like the "per-table" bit should be part of the schema, not decided locally based on incomplete info.
> 
> OK. Any place you have in mind? Part of `with replication = { ... }` set doesn't work, all field in it [are treated as replication factor integers](https://github.com/scylladb/scylladb/pull/16364#discussion_r1422569803)

We have scylla specific keyspace for our extensions: "scylla_keyspaces". IIRC there was a similar problem to what you describe in the link when we added in_memory option to tables (describe table failed)  and we solved by putting the option into scylla specific "scylla_tables".

xemul replied:
@tgrabiec , @gleb-cloudius , FYI, keeping this bit in system_schema.scylla_keyspaces steps on the same issue -- when merging schema this place

```c++
future<lw_shared_ptr<query::result_set>> extract_scylla_specific_keyspace_info(distributed<service::storage_proxy>& proxy, const schema_result_value_type& partition) {
    lw_shared_ptr<query::result_set> scylla_specific_rs;
    if (proxy.local().features().cluster_schema_features().contains<schema_feature::SCYLLA_KEYSPACES>()) {
        auto&& rs = partition.second;
        if (rs->empty()) {
            co_await coroutine::return_exception(std::runtime_error("query result has no rows"));
        }
        auto&& row = rs->row(0);
        auto keyspace_name = row.get_nonnull<sstring>("keyspace_name");
        auto keyspace_key = dht::decorate_key(*scylla_keyspaces(), partition_key::from_singular(*scylla_keyspaces(), keyspace_name));
        scylla_specific_rs = co_await db::system_keyspace::query(proxy.local().get_db(), NAME, SCYLLA_KEYSPACES, keyspace_key);
    }
    co_return scylla_specific_rs;
}
```

doesn't load anything from scylla_keyspaces because feature is not enabled yet

tgrabiec replied:
Perhaps it should always extract (supported = may be there, enabled = allow putting it in).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5176/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6567/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5176/)
#### Failed Tests (36/22265):
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_topology_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_topology_changes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_changes)
- [test_streaming_is_guarded_by_topology_guard](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_streaming_is_guarded_by_topology_guard) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_streaming_is_guarded_by_topology_guard)
- [test_topology_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_topology_changes_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_changes)
- [test_streaming_is_guarded_by_topology_guard](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_streaming_is_guarded_by_topology_guard_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_streaming_is_guarded_by_topology_guard)
- [test_topology_changes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_topology_changes_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_changes)
- [test_streaming_is_guarded_by_topology_guard](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_streaming_is_guarded_by_topology_guard_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_streaming_is_guarded_by_topology_guard)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5176/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
#### Build Details:
- Duration: 2 hr 54 min
- Builder: spider8.cloudius-systems.com

nyh commented:
> After this patch it's still possible to pass the `initial_tables` into underlying keyspace creation. The only difference is that this option becomes an _option_, i.e. -- when not specified, its value will be auto-estimated, but tables will be turned on anyway

Yes, I know, except that now if Alternator users don't specify anything special (but enable the tablets experimental feature globally), tablets will be used by default on every table, which I thought I didn't want (because of missing CDC) but maybe, actually, it isn't such a bad idea. And in any case whatever you name the new option (initial_tablets? tablets?) I'll need to allow changing from Alternator as well. So I suggest you guys make a decision on this issue first and then I'll send a new version of https://github.com/scylladb/scylladb/pull/16353 that deals with that decision.

xemul commented:
upd:
- move test_tablet_drain_failure_during_decommission to topology_custom suite
- merge topology changes before schema when applying raft snapshot

@tgrabiec your [suggestion](https://github.com/scylladb/scylladb/pull/16364#discussion_r1422821841) to keep "this keyspace uses tablets" bit in schema is not yet here, since the main goal of this PR is to make it possible to run sct perf tests, so I thought to make that change as a followup. If you think it should be done instantly, let me know, I'll revisit.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5239/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6641/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5239/)
#### Failed Tests (6/23454):
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5239/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5239/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5239/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [topology_custom.test_topology_failure_recovery.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5239/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_topology_failure_recovery_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_failure_recovery.release.1)
- [topology_custom.test_topology_failure_recovery.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5239/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_topology_failure_recovery_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_failure_recovery.release.2)
- [topology_custom.test_topology_failure_recovery.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5239/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_topology_failure_recovery_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_failure_recovery.release.3)
#### Build Details:
- Duration: 2 hr 20 min
- Builder: spider6.cloudius-systems.com

xemul commented:
upd:
- add handling of `initial_tablets: 0` option -- when a table is created it auto-calculates some better value
- when the tablets feature is ON and initial_tablets is not specified, its zero value is assigned on user's behalf thus always enabling tablets
- unit test for default configuration

In the cover letter the issue is referenced, not fixed. The current plan, as I see it, is to get rid of per-keyspace initial_tablets and replace it with boolean switch bit and per-table opt-out one

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5267/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5267/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6669/)
#### Failed Tests (6/23460):
- [topology_custom.test_tablets.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5267/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_tablets_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_tablets.release.1)
- [topology_custom.test_tablets.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5267/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_tablets_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_tablets.release.2)
- [topology_custom.test_tablets.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5267/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_tablets_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_tablets.release.3)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5267/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5267/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5267/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
#### Build Details:
- Duration: 2 hr 21 min
- Builder: spider5.cloudius-systems.com

xemul commented:
upd:
- test_topology_failure_recovery.py must be skipped in release mode

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5270/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6675/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5270/)
#### Build Details:
- Duration: 5 hr 46 min
- Builder: monster

xemul commented:
Running SCT tests with this patch steps on

```
2023-12-19T10:26:15.413+00:00 perf-regression-ubuntu-db-node-5c2c1f96-1     !INFO | scylla[12358]:  [shard 0:strm] schema_tables - Altering keyspace system_auth
2023-12-19T10:26:15.413+00:00 perf-regression-ubuntu-db-node-5c2c1f96-1     !INFO | scylla[12358]:  [shard 0:strm] migration_manager - Gossiping my schema version 09dbe732-9e59-11ee-5e68-8dd69330ab46
2023-12-19T10:26:15.413+00:00 perf-regression-ubuntu-db-node-5c2c1f96-1     !INFO | scylla[12358]:  [shard 0:strm] schema_tables - Schema version changed to 09dbe732-9e59-11ee-5e68-8dd69330ab46
2023-12-19T10:26:18.663+00:00 perf-regression-ubuntu-db-node-5c2c1f96-1      !ERR | scylla[12358]:  [shard 0:strm] database - Tried to obtain per-keyspace effective replication map of system_auth but it's per-table, at: 0x60dd8be 0x60dde8
0 0x60de158 0x5ba7657 0x1a0b4b6 0x3e776a4 0x3e63356 0x433df82 0x437a661 0x437ac3b 0x5bd8bef 0x5bd9ec7 0x5bd9239 0x5b79e27 0x5b78fdc 0x13196f9 0x131b150 0x1317d7c /opt/scylladb/libreloc/libc.so.6+0x27b89 /opt/scylladb/libreloc/libc.so.6+0x
27c4a 0x13157a4
   --------
...
Stack trace of thread 12358:
#0  0x00007ff608f3f884 __pthread_kill_implementation (libc.so.6 + 0x8e884)
#1  0x00007ff608eeeafe raise (libc.so.6 + 0x3dafe)
#2  0x00007ff608ed787f abort (libc.so.6 + 0x2687f)
#3  0x0000000005ba76d8 _ZN7seastar17on_internal_errorERNS_6loggerESt17basic_string_viewIcSt11char_traitsIcEE (scylla + 0x59a76d8)
#4  0x0000000001a0b4b7 _ZNK7replica8keyspace29get_effective_replication_mapEv (scylla + 0x180b4b7)
#5  0x0000000003e776a5 _ZZN7locator38global_vnode_effective_replication_map17get_keyspace_ermsERN7seastar7shardedIN7replica8databaseEEESt17basic_string_viewIcSt11char_traitsIcEEENK3$_0clERS4_ (scylla + 0x3c776a5)
#6  0x0000000003e63357 _ZN7locator37make_global_effective_replication_mapERN7seastar7shardedIN7replica8databaseEEESt17basic_string_viewIcSt11char_traitsIcEE (scylla + 0x3c63357)
#7  0x000000000433df83 _ZN14repair_service15do_repair_startEN7seastar13basic_sstringIcjLj15ELb1EEESt13unordered_mapIS2_S2_St4hashIS2_ESt8equal_toIS2_ESaISt4pairIKS2_S2_EEE (scylla + 0x413df83)
#8  0x000000000437a662 _ZN7seastar8futurizeINS_6futureIiEEE6invokeIRZNS_7shardedI14repair_serviceE9invoke_onIZ12repair_startRS7_NS_13basic_sstringIcjLj15ELb1EEESt13unordered_mapISB_SB_St4hashISB_ESt8equal_toISB_ESaISt4pairIKSB_SB_EEEE3$_0
JES2_EET1_jNS_21smp_submit_to_optionsEOT_DpOT0_EUlvE_JEEES2_SQ_ST_ (scylla + 0x417a662)
#9  0x000000000437ac3c _ZN7seastar17smp_message_queue15async_work_itemIZNS_7shardedI14repair_serviceE9invoke_onIZ12repair_startRS4_NS_13basic_sstringIcjLj15ELb1EEESt13unordered_mapIS8_S8_St4hashIS8_ESt8equal_toIS8_ESaISt4pairIKS8_S8_EEEE3
$_0JENS_6futureIiEEEET1_jNS_21smp_submit_to_optionsEOT_DpOT0_EUlvE_E15run_and_disposeEv (scylla + 0x417ac3c)
#10 0x0000000005bd8bf0 _ZN7seastar7reactor14run_some_tasksEv (scylla + 0x59d8bf0)
#11 0x0000000005bd9ec8 _ZN7seastar7reactor6do_runEv (scylla + 0x59d9ec8)
#12 0x0000000005bd923a _ZN7seastar7reactor3runEv (scylla + 0x59d923a)
#13 0x0000000005b79e28 _ZN7seastar12app_template14run_deprecatedEiPPcOSt8functionIFvvEE (scylla + 0x5979e28)
#14 0x0000000005b78fdd _ZN7seastar12app_template3runEiPPcOSt8functionIFNS_6futureIiEEvEE (scylla + 0x5978fdd)
#15 0x00000000013196fa _ZL11scylla_mainiPPc (scylla + 0x11196fa)
#16 0x000000000131b151 _ZNKSt8functionIFiiPPcEEclEiS1_ (scylla + 0x111b151)
#17 0x0000000001317d7d main (scylla + 0x1117d7d)
#18 0x00007ff608ed8b8a __libc_start_call_main (libc.so.6 + 0x27b8a)
#19 0x00007ff608ed8c4b __libc_start_main@@GLIBC_2.34 (libc.so.6 + 0x27c4b)
#20 0x00000000013157a5 _start (scylla + 0x11157a5)
```

https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/pavel/job/perf-regression-throughput-sample-try/7/consoleText

xemul commented:
Previously the system_auth was created with
```
2023-12-19T10:18:44.176+00:00 perf-regression-ubuntu-db-node-5c2c1f96-1     !INFO | scylla[12358]:  [shard 0:main] migration_manager - Create new Keyspace: KSMetaData{name=system_auth, strategyClass=org.apache.cassandra.locator.SimpleStra
tegy, strategyOptions={replication_factor=1}, cfMetaData={}, durable_writes=1, userTypes=org.apache.cassandra.config.UTMetaData@0x60000911b0a0}
```

avikivity commented:
> > After this patch it's still possible to pass the `initial_tables` into underlying keyspace creation. The only difference is that this option becomes an _option_, i.e. -- when not specified, its value will be auto-estimated, but tables will be turned on anyway
> 
> Yes, I know, except that now if Alternator users don't specify anything special (but enable the tablets experimental feature globally), tablets will be used by default on every table, which I thought I didn't want (because of missing CDC) but maybe, actually, it isn't such a bad idea.

People who turn on experimental flags are responsible for understanding the implications.


> And in any case whatever you name the new option (initial_tablets? tablets?) I'll need to allow changing from Alternator as well. So I suggest you guys make a decision on this issue first and then I'll send a new version of #16353 that deals with that decision.

Best is if it's on by default when tablets are enabled, so we don't have to adjust all tests.

xemul commented:
> Back then repair started by mistake and was fixed

I mean this comment https://github.com/scylladb/scylladb/pull/16364#discussion_r1425388543

xemul commented:
Test deliberately runs repair on it after changing the replication strategy to network-topology one (originally it's created with simple strategy):

```python
        NetworkTopologyReplicationStrategy(**datacenters).apply(node, "system_auth")
        res = node.run_cqlsh('DESC KEYSPACE system_auth',
                             num_retry_on_failure=3)
        self.log.debug("system_auth description: %s", res.stdout)
        self.log.info('repair system_auth keyspace ...')
        for node in self.db_cluster.nodes:
            node.run_nodetool(sub_cmd="repair -pr system_auth", timeout=MINUTE_IN_SEC * 20)
        self.log.info('repair system_auth keyspace done')
```
https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/tester.py#L985-L992

@tgrabiec , what's the correct way to change replication strategy with tablets nowadays?

xemul commented:
OK, so just changing the keyspace strategy works, but pre-existing tables continue to be vnode-based (at least immediately after ks altering)

```
    @pytest.mark.asyncio
    async def test_tablet_change_replication_strategy(manager: ManagerClient):
        cfg = {'enable_user_defined_functions': False,
               'experimental_features': ['tablets', 'consistent-topology-changes']}
        server = await manager.server_add(config=cfg)
    
        cql = manager.get_cql()
        await cql.run_async("CREATE KEYSPACE test WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};")
        await cql.run_async("CREATE TABLE test.test (pk int PRIMARY KEY, c int);")
        await cql.run_async("ALTER KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy'};")
    
        res = await cql.run_async("DESCRIBE KEYSPACE test")
        assert "'initial_tablets': '0'" in res[0].create_statement, "initial_tablets not configured on keyspace"
    
        res = await cql.run_async("SELECT * FROM system.tablets")
        for row in res:
            if row.keyspace_name == 'test' and row.table_name == 'test':
                assert row.tablet_count > 0, "zero tablets allocated"
                break
        else:
>           assert False, "tablets not allocated"
E           AssertionError: tablets not allocated
E           assert False
```

tgrabiec commented:
> @tgrabiec , what's the correct way to change replication strategy with tablets nowadays?

We don't support altering keyspace from vnode-based to tablet-based, we should refuse to alter in this case.

xemul commented:
upd:
- when ALTER-ing keyspace preserve its existing vnode/tablets state, turn tablets ON automatically only if the feature is enabled

With that SCT performance regression test finally passes:
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/pavel/job/perf-regression-throughput-sample-try/8/

tgrabiec commented:
How does the user opt-out of tablets on per-keyspace basis?

xemul commented:
> How does the user opt-out of tablets on per-keyspace basis?

With this PR it's not possible, we can't use non-integer parameters for NetworkTopology strategy

xemul commented:
> With this PR it's not possible

Well... not quite

```
CREATE KEYSPACE ks WITH REPLICATION = {'class': 'SimpleStrategy', ... };
ALTER KEYSPACE ks WITH = {'class': 'NetworkTopologyStrategy', ...};
```

would do it, though it's not extremely elegant

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5334/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6751/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5334/)
#### Build Details:
- Duration: 2 hr 34 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
@xemul please rebase and fix any remaining review comments

xemul commented:
> @xemul please rebase and fix any remaining review comments

Sure thing!

xemul commented:
upd:
- rebased
- changed CREATE KEYSPACE syntax to (likely) more convenient one (see updated PR description for details)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5683/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7079/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5683/)
#### Build Details:
- Duration: 1 hr 53 min
- Builder: spider1.cloudius-systems.com

xemul commented on cql3/statements/ks_prop_defs.cc:
```diff
@@ -83,7 +77,7 @@ void ks_prop_defs::validate() {
     }
 
-    static std::set<sstring> keywords({ sstring(KW_DURABLE_WRITES), sstring(KW_REPLICATION), sstring(KW_STORAGE) });
+    static std::set<sstring> keywords({ sstring(KW_DURABLE_WRITES), sstring(KW_REPLICATION), sstring(KW_STORAGE), sstring(KW_TABLETS) });
```
There was #16193 to document the thing, but it looks like it was merged so yes, the text needs updating

xemul commented on cql3/statements/ks_prop_defs.cc:
```diff
@@ -114,13 +108,43 @@ data_dictionary::storage_options ks_prop_defs::get_storage_options() const {
 }
 
+std::optional<unsigned> ks_prop_defs::get_initial_tablets(const sstring& strategy_class) const {
+    // FIXME -- this should be ignored somehow else
```
The "problem" is that if you add `TABLETS = { ... }` when creating a ks with, say, SimpleStrategy it will won't cause any errors, replication strategy will remain vnode-based, _but_ `system_schema.scylla_keyspaces` will be updated with `initial_tablets` value which doesn't look correct. So there are two options here

1. bail out with error if `CREATE KEYSPACE` asks for `TABLETS` _AND_ for `REPLICATION` with class that doesn't support tablets
2. ignore `TABLETS` from `CREATE KEYSPACE` in case `REPLICATION`'s class doesn't support tablets

I picked the latter. WDYT?

xemul commented on cql3/statements/ks_prop_defs.cc:
```diff
@@ -114,13 +108,43 @@ data_dictionary::storage_options ks_prop_defs::get_storage_options() const {
+    }
+
+    if (!tablets_options->empty()) {
+        throw exceptions::configuration_exception(sstring("Unrecognized tablets options ") + tablets_options->begin()->first);
```
Good point

xemul commented on cql3/statements/ks_prop_defs.cc:
```diff
@@ -121,7 +121,21 @@ std::optional<unsigned> ks_prop_defs::get_initial_tablets(const sstring& strateg
+        auto enabled = it->second;
+        tablets_options->erase(it);
+
+        if (enabled == "true") {
```
> Or everything is converted to strings?

The options are `std::map<sstring, sstring>` so its all strings up until the code that interprets them (this code)

> Are the strings always lowercase?

They are the way user passes them. Then it's up to the parser code whether or not to care about case. E.g.  `CREATE TABLE ... CACHING = { 'enabled' = false }` parser checks for `option.second == 'true'` i.e. -- doesn't care. Opposite to that `CREATE KEYSPACE ... DURABLE_WRITES = false` does care and even accepts `1` or `yes` as true.

xemul commented on cql3/statements/ks_prop_defs.cc:
```diff
@@ -121,7 +121,21 @@ std::optional<unsigned> ks_prop_defs::get_initial_tablets(const sstring& strateg
+        if (enabled == "true") {
+            ret = 0; // even if 'initial' is not set, it'll start with auto-detection
+        } else if (enabled == "false") {
+            goto skip;
```
> I think there's a problem here: "enabled=false, initial_tablets=5" will complain that "unrecognized tablets option initial_tablets" because you "skip" over the code below that removes the initial_tablets.

That's deliberate. If user doesn't enable tablets all other options become unrecognized.

>  This will be wrong, and confusing for any user who'll see this message.

What do you suggest? Ignore other options or report back some other message?

bhalevy replied:
We have a similar use case with compaction options.
In this case we accept all options, including 'enabled', which can be enabled/disabled later on with yet another ALTER TABLE query, while the other options are persisted.

xemul replied:
It might not work for initial_tablets. There's no dedicated on/off switch on keyspace_metadata nor there's one in system_schema.scylla_keyspaces table. Only the initial_tablets value and if it's present, tablets are ON, if it's missing tablets are OFF

xemul commented on test/topology_custom/test_tablets.py:
```diff
@@ -48,10 +48,10 @@ async def test_tablet_default_initialization(manager: ManagerClient):
 
     res = await cql.run_async("SELECT * FROM system_schema.scylla_keyspaces WHERE keyspace_name = 'test'")
-    assert res[0].initial_tablets > 0, "initial_tablets not configured"
+    assert res[0].initial_tablets == 0, "initial_tablets not configured"
```
>  It's still not clear why you test for initial_tablets=0 and not initial_tablets>0.

Because if you start with default tablets configuration, it will be 0 initial tablets _in_ _system_schema_ _table_. When a table is created for a keyspace that has 0 initial tablets it will [auto-estimate](https://github.com/scylladb/scylladb/pull/16364#discussion_r1448961336) the real number.

xemul commented on test/boost/tablets_test.cc:
```diff
@@ -31,7 +31,7 @@ using namespace replica;
 
 static api::timestamp_type current_timestamp(cql_test_env& e) {
-    return api::new_timestamp();
+    return utils::UUID_gen::micros_timestamp(e.get_system_keyspace().local().get_last_group0_state_id().get0()) + 1;
```
This test tries to mutate entries in tablets table. When it does so there are some pre-existing mutations in this table and in order for the test mutation to apply successfully their timestamps need to be "later" than those sitting in the table. "Those sitting in the table" got their timestamps from group0, so this is the way to get a "later" timestamp.

xemul commented:
upd:
- rebased
- fixed ddl.rst documentation
- renamed `estimate_initial_tablets()` and added a description of what it does
- fixed "bad tablets option" exception text
- added a comment to test describing current_timestamp() magic

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5735/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7111/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5735/)
#### Build Details:
- Duration: 1 hr 55 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
I think the UX would be nicer if we ignore the other option(s) when enabled=false, rather than err on them, if that's not too complicated.

nyh commented:
> I think the UX would be nicer if we ignore the other option(s) when enabled=false, rather than err on them, if that's not too complicated.

I think so too: I think that both it will be more convenient (you can quickly disable by setting enabled=false and leaving the other option intact), and less misleading: It's misleading to get an error message like "option initial_tablets not supported", while this option is supported - not just in combination with the other options you tried to use.

xemul commented:
upd:
- rebased
- ignore all TABLETS map options in case `'enabled' = false` is met

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5777/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7146/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5777/)
#### Failed Tests (1/23781):
- [test_cql_query_filtering_without_indexes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5777/testReport/junit/cql_tests/TestTruncate/Tests___dtest___test_cql_query_filtering_without_indexes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cql_query_filtering_without_indexes)
#### Build Details:
- Duration: 1 hr 35 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5780/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7149/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5780/)
#### Build Details:
- Duration: 1 hr 29 min
- Builder: spider1.cloudius-systems.com

tzach commented:
@annastuchlik please note the doc update. We should link to it from the (future) Tablets page.
